ls1intum/Artemis
# `Development`: Fix flakiness in cypress e2e tests for programming exercise assessments

### Motivation and Context


The `ProgrammingExerciseAssessment` spec sometimes fails. It seems like the student sometimes does not get the additional feedback from the tutor even though the assessment has been saved properly. However this only happens sometimes when the Cypress tests are executed in a build agent. I could never reproduce this issue locally (running the tests locally against TS3 where the build agent failed the test) and also couldn't reproduce it manually. I believe the issues arises from updating the due date and assessment due date of the programming exercise dynamically during test execution. This is does not represent a very realistic use case in the first place, so we should improve this.

### Description

In this PR I merged the 3 existing tests, which were dependent on each other, into one larger test. I also removed the code to set the due date and assessment due date via direct request to the Artemis Server and instead set the dates directly during creation of the programming exercise. This should also make the test more realistic because usually the dates are set in the beginning of an exercise and aren't set/changed on the fly. This does however slightly reduce the performance of the test because it now waits until the set deadline is in the past instead of just updating the due date. The test runs 1 minute and 13 seconds on average, which is still fine in my opinion.

### Steps for Testing

Since the edited test is flaky I ran that specific test several times over 3 build plan executions.

1. [Executed the normal test suite once (all tests)](https://bamboo.ase.in.tum.de/browse/ARTEMIS-AETBBT-714)
2. [Executed the `ProgrammingExerciseAssessment` once](https://bamboo.ase.in.tum.de/browse/ARTEMIS-AETBBT-715)
3. [Executed the `ProgrammingExerciseAssessment` 5 times](https://bamboo.ase.in.tum.de/browse/ARTEMIS-AETBBT-715)
4. [Executed the `ProgrammingExerciseAssessment` 10 times](https://bamboo.ase.in.tum.de/browse/ARTEMIS-AETBBT-716)

To see the results properly one has to look at the logs of the build plan. In total the `ProgrammingExerciseAssessment` has been executed 17 times and always passed. I can't say for sure that the flakiness has been fixed, but the results look good to me.

### Review Progress





#### Code Review
